// ng-modules/dist/providers/modal/modal.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Injectable } from '@angular/core';
import { ModalComponent, AlertComponent, ConfirmComponent, LoadingComponent, ToastComponent, InfoConfirmComponent } from '../../components';
import { UtilProvider } from '../util/util';
import { I18nProvider } from '../i18n/i18n';
var ModalProvider = (function () {
    function ModalProvider(util, i18n) {
        this.util = util;
        this.i18n = i18n;
    }
    ModalProvider.prototype.alert = function (options) {
        return this.modal(AlertComponent, options, this.util.extend({
            title: this.i18n.get("__8"),
            width: 500,
            height: 315,
        }, options, true));
    };
    ModalProvider.prototype.confirm = function (options) {
        return this.modal(ConfirmComponent, options, this.util.extend({
            title: this.i18n.get("__8"),
            width: 500,
            height: 315,
        }, options, true));
    };
    ModalProvider.prototype.inforConfirm = function (options) {
        if (typeof options == "string") {
            options = {
                content: options,
            };
        }
        options.css = options.css || "content-left";
        return this.modal(InfoConfirmComponent, options, this.util.extend({
            width: 500,
            height: 200,
        }, options, true));
    };
    ModalProvider.prototype.loading = function (content, options) {
        return this.modal(LoadingComponent, {
            content: content,
        }, this.util.extend({
            width: content ? 160 : 110,
            height: content ? 160 : 110,
        }, options, true));
    };
    ModalProvider.prototype.modal = function (component, data, options) {
        return ModalComponent.Instance.create(component, data, options);
    };
    ModalProvider.prototype.toast = function (content) {
        if (typeof content == "string") {
            content = { content: content, };
        }
        ToastComponent.Instance.create(content);
    };
    ModalProvider = __decorate([
        Injectable(),
        __metadata("design:paramtypes", [UtilProvider,
            I18nProvider])
    ], ModalProvider);
    return ModalProvider;
}());
export { ModalProvider };
