// ng-modules/dist/module.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { NgModule } from '@angular/core';
import { ListComponent, ListCellComponent, ModalComponent, ModalContentComponent, AlertComponent, MoveableComponent, FullscreenComponent, ConfirmComponent, FormInputStatus, SelectorComponent, LoadingComponent, ToastComponent, TabsComponent, TabContentDirective, TabSwitchDirective, UEditorComponent, DaterangeDirective, DatetimeDirective, InfoConfirmComponent, PopoverComponent } from './components';
import { ModalProvider, DomProvider, UtilProvider, AnimationProvider, I18nProvider, I18nProviderFactory, PopoverProvider, I18NToken, } from './providers';
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { Datetime, SeHtml, SeStyle, SeUrl, I18NPipe } from './pipes';
import { RouterModule } from '@angular/router';
var CoreModule = (function () {
    function CoreModule() {
    }
    CoreModule_1 = CoreModule;
    CoreModule.forRoot = function (config) {
        if (config === void 0) { config = {}; }
        return {
            ngModule: CoreModule_1,
            providers: [
                ModalProvider,
                PopoverProvider,
                DomProvider,
                UtilProvider,
                AnimationProvider,
                { provide: I18NToken, useValue: config.i18n || {}, },
                { provide: I18nProvider, useFactory: I18nProviderFactory, deps: [I18NToken], },
            ]
        };
    };
    var CoreModule_1;
    CoreModule = CoreModule_1 = __decorate([
        NgModule({
            declarations: [
                ListComponent,
                ListCellComponent,
                ModalComponent,
                ModalContentComponent,
                Datetime,
                I18NPipe,
                SeHtml,
                SeStyle,
                SeUrl,
                AlertComponent,
                ConfirmComponent,
                InfoConfirmComponent,
                LoadingComponent,
                MoveableComponent,
                FullscreenComponent,
                FormInputStatus,
                SelectorComponent,
                ToastComponent,
                TabsComponent,
                TabContentDirective,
                TabSwitchDirective,
                UEditorComponent,
                DaterangeDirective,
                DatetimeDirective,
                PopoverComponent,
            ],
            entryComponents: [
                ListComponent,
                ModalComponent,
                AlertComponent,
                ConfirmComponent,
                InfoConfirmComponent,
                FullscreenComponent,
                LoadingComponent,
                ToastComponent,
                TabsComponent,
                UEditorComponent,
                PopoverComponent,
            ],
            exports: [
                ListComponent,
                ListCellComponent,
                ModalComponent,
                ModalContentComponent,
                Datetime,
                I18NPipe,
                SeHtml,
                SeStyle,
                SeUrl,
                AlertComponent,
                ConfirmComponent,
                InfoConfirmComponent,
                LoadingComponent,
                RouterModule,
                MoveableComponent,
                FullscreenComponent,
                FormInputStatus,
                SelectorComponent,
                ToastComponent,
                TabsComponent,
                TabContentDirective,
                TabSwitchDirective,
                UEditorComponent,
                DaterangeDirective,
                DatetimeDirective,
                PopoverComponent,
            ],
            imports: [
                CommonModule,
                FormsModule,
                RouterModule,
                ReactiveFormsModule,
            ],
        }),
        __metadata("design:paramtypes", [])
    ], CoreModule);
    return CoreModule;
}());
export { CoreModule };
