// ng-modules/dist/pipes/se/se.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Pipe } from "@angular/core";
import { DomSanitizer } from '@angular/platform-browser';
var SeUrl = (function () {
    function SeUrl(domSanitizer) {
        this.domSanitizer = domSanitizer;
    }
    SeUrl.prototype.transform = function (value) {
        return this.domSanitizer.bypassSecurityTrustUrl(value);
    };
    SeUrl = __decorate([
        Pipe({
            name: "SeUrl"
        }),
        __metadata("design:paramtypes", [DomSanitizer])
    ], SeUrl);
    return SeUrl;
}());
export { SeUrl };
var SeStyle = (function () {
    function SeStyle(domSanitizer) {
        this.domSanitizer = domSanitizer;
    }
    SeStyle.prototype.transform = function (value) {
        return this.domSanitizer.bypassSecurityTrustStyle(value);
    };
    SeStyle = __decorate([
        Pipe({
            name: "SeStyle"
        }),
        __metadata("design:paramtypes", [DomSanitizer])
    ], SeStyle);
    return SeStyle;
}());
export { SeStyle };
var SeHtml = (function () {
    function SeHtml(domSanitizer) {
        this.domSanitizer = domSanitizer;
    }
    SeHtml.prototype.transform = function (value) {
        return this.domSanitizer.bypassSecurityTrustHtml(value);
    };
    SeHtml = __decorate([
        Pipe({
            name: "SeHtml"
        }),
        __metadata("design:paramtypes", [DomSanitizer])
    ], SeHtml);
    return SeHtml;
}());
export { SeHtml };
