// ng-modules/dist/components/form/datetime/datetime.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Directive, ElementRef, forwardRef } from "@angular/core";
import { NG_VALUE_ACCESSOR } from "@angular/forms";
import { Input } from "@angular/core";
import { UtilProvider } from "../../../providers/util/util";
import { I18nProvider } from "../../../providers/i18n/i18n";
var DatetimeDirective = (function () {
    function DatetimeDirective(element, util) {
        this.element = element;
        this.util = util;
    }
    DatetimeDirective_1 = DatetimeDirective;
    DatetimeDirective.prototype.writeValue = function (value) {
        if (value) {
            $(this.element.nativeElement).data("daterangepicker").setStartDate(value);
        }
    };
    DatetimeDirective.prototype._registerOnChange = function () { };
    DatetimeDirective.prototype.registerOnChange = function (fn) {
        this._registerOnChange = fn;
    };
    DatetimeDirective.prototype.touched = function () { };
    DatetimeDirective.prototype.registerOnTouched = function (fn) {
        this.touched = fn;
    };
    DatetimeDirective.prototype.setDisabledState = function (isDisabled) { };
    Object.defineProperty(DatetimeDirective.prototype, "datetime", {
        get: function () {
            return this._config;
        },
        set: function (value) {
            this._config = value;
            this.setting(value);
        },
        enumerable: true,
        configurable: true
    });
    ;
    DatetimeDirective.prototype.ngOnInit = function () {
    };
    DatetimeDirective.prototype.setting = function (config) {
        var data = this.util.extend({
            singleDatePicker: true,
            autoUpdateInput: false,
            locale: I18nProvider.get("dateSetting"),
        }, config, true);
        $(this.element.nativeElement).daterangepicker(data);
        var that = this;
        $(this.element.nativeElement).on('apply.daterangepicker', function (ev, picker) {
            var format = "YYYY-MM-DD";
            if (that.datetime && that.datetime.locale && that.datetime.locale.format) {
                format = that.datetime.locale.format;
            }
            that._registerOnChange();
            $(this).val(picker.startDate.format(format));
        });
        $(this.element.nativeElement).on('cancel.daterangepicker', function () {
        });
    };
    var DatetimeDirective_1;
    __decorate([
        Input("datetime"),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], DatetimeDirective.prototype, "datetime", null);
    DatetimeDirective = DatetimeDirective_1 = __decorate([
        Directive({
            selector: "[datetime]",
            providers: [
                { provide: NG_VALUE_ACCESSOR, useExisting: forwardRef(function () { return DatetimeDirective_1; }), multi: true }
            ],
        }),
        __metadata("design:paramtypes", [ElementRef,
            UtilProvider])
    ], DatetimeDirective);
    return DatetimeDirective;
}());
export { DatetimeDirective };
