// ng-modules/dist/components/popover/popover.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Component, Input, TemplateRef, ViewChild, ElementRef, HostListener } from "@angular/core";
import { PopoverProvider } from "../../providers/modal/popover";
var PopoverComponent = (function () {
    function PopoverComponent(ele, popoverProvider) {
        this.ele = ele;
        this.popoverProvider = popoverProvider;
        this.position = "right";
        this._popEleMouseOver = false;
    }
    Object.defineProperty(PopoverComponent.prototype, "popover", {
        set: function (value) {
            if (value instanceof TemplateRef) {
                this.tpl = value;
                this._popover = this.tpl;
            }
        },
        enumerable: true,
        configurable: true
    });
    ;
    PopoverComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.popCtrl = this.popoverProvider.create({
            content: this.popEle.nativeElement,
            position: this.position,
            width: this.width,
            height: this.height,
            parent: this.ele.nativeElement,
        });
        this.popEle.nativeElement.addEventListener("mouseover", function () {
            _this._popEleMouseOver = true;
            _this.popCtrl.show();
            setTimeout(function () {
                _this._popEleMouseOver = false;
            }, 100);
        });
        this.popEle.nativeElement.addEventListener("mouseout", function () {
            _this.popCtrl.hide();
        });
    };
    PopoverComponent.prototype.bindBodyClick = function () {
        var _this = this;
        var popClick = false;
        var popClickTimer;
        this.popEle.nativeElement.addEventListener("click", this._popFn = function (e) {
            popClick = true;
            clearTimeout(popClickTimer);
            popClickTimer = setTimeout(function () {
                popClick = false;
            }, 300);
        });
        document.body.addEventListener("click", this._fn = function (e) {
            if (popClick)
                return;
            if (!e.path.some(function (item) {
                if (item == _this.popoverOutbox.nativeElement || item == _this.parentEle) {
                    return true;
                }
            })) {
                _this.popCtrl.hide();
            }
        });
    };
    PopoverComponent.prototype.clearBodyClick = function () {
        document.body.removeEventListener("click", this._fn);
    };
    PopoverComponent.prototype.mouseOver = function (e) {
        this.popCtrl.show();
    };
    PopoverComponent.prototype.mouseOut = function (e) {
        var _this = this;
        setTimeout(function () {
            if (!_this._popEleMouseOver) {
                _this.popCtrl.hide();
            }
        }, 50);
    };
    PopoverComponent.prototype.getRect = function () {
        var pop = this.parentEle;
        return pop.getBoundingClientRect();
    };
    __decorate([
        Input("popover"),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], PopoverComponent.prototype, "popover", null);
    __decorate([
        ViewChild("popEle"),
        __metadata("design:type", ElementRef)
    ], PopoverComponent.prototype, "popEle", void 0);
    __decorate([
        ViewChild("popoverOutbox"),
        __metadata("design:type", ElementRef)
    ], PopoverComponent.prototype, "popoverOutbox", void 0);
    __decorate([
        Input("popWidth"),
        __metadata("design:type", Object)
    ], PopoverComponent.prototype, "width", void 0);
    __decorate([
        Input("popHeight"),
        __metadata("design:type", Object)
    ], PopoverComponent.prototype, "height", void 0);
    __decorate([
        Input("popPosition"),
        __metadata("design:type", Object)
    ], PopoverComponent.prototype, "position", void 0);
    __decorate([
        HostListener('mouseover', ['$event']),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", void 0)
    ], PopoverComponent.prototype, "mouseOver", null);
    __decorate([
        HostListener('mouseout', ['$event']),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", void 0)
    ], PopoverComponent.prototype, "mouseOut", null);
    PopoverComponent = __decorate([
        Component({
            selector: "[popover]",
            template: "<ng-content></ng-content> <div #popEle> <ng-container *ngTemplateOutlet=\"tpl\"></ng-container> </div> ",
        }),
        __metadata("design:paramtypes", [ElementRef,
            PopoverProvider])
    ], PopoverComponent);
    return PopoverComponent;
}());
export { PopoverComponent };
