// ng-modules/src/components/modal/confirm/confirm.ts
import { Component, TemplateRef } from "@angular/core";
import { ModalController } from "../../../providers/modal/modalController";
import { isBoolean } from "util";
import { ModalOptions } from "../modal";
import { FormBuilder } from "@angular/forms";
import { I18nProvider } from "../../../providers/i18n/i18n";

export interface ConfirmOptions extends ModalOptions {
    title?: string;
    image?: string;
    content?: string;
    hideOk?: boolean;
    okText?: string;
    beforeOk?: (data: any) => boolean | Promise<boolean>;
    hideCancel?: boolean;
    cancelText?: string;
    beforeCancel?: (data: any) => boolean | Promise<boolean>;
    buttons?: ConfirmButton[];
    inputs?: ConfirmInput[];
    buttonsCss?: string;
    contentTpl?: TemplateRef<any>;
}

export interface ConfirmButton {
    text: string;
    cssClass?: string;
    canClose?: (data: any) => boolean | Promise<boolean>;
}

export interface ConfirmInput {
    /**输入框字段 */
    name: string;
    /**输入框类型，默认text，同 <input /> 类型 */
    type?: string;
    placeholder?: string;
    /**默认值 */
    value?: string;
    label?: string;
    /**选择框option */
    options?: { value: any, name: string }[];
}

@Component({
    templateUrl: "./confirm.html",
})
export class ConfirmComponent {
    data;
    form;
    constructor(
        private modalController: ModalController,
        private formBuilder: FormBuilder,
        private i18n: I18nProvider,
    ) { }
    content: string = "";
    buttons: ConfirmButton[];
    inputs: ConfirmInput[];
    buttonsCss = "";
    contentTpl: TemplateRef<any>;

    clicking: boolean;
    async buttonClick(btn: (ConfirmButton & { type?: "ok" | "cancel" })) {
        this.clicking = true;
        try {
            let result;
            if (btn.canClose) {
                result = await btn.canClose(this.data);
            }
            if (typeof result == "undefined") {
                if (btn.type == "ok") {
                    this.modalController.sendMessage(this.data);
                    this.modalController.close();
                } else if (btn.type == "cancel") {
                    this.modalController.sendError(this.data);
                    this.modalController.close();
                } else {
                    this.modalController.sendMessage(this.data || btn);
                    this.modalController.close();
                }
            } else {
                if (result) {
                    this.modalController.sendMessage(this.data || result);
                    this.modalController.close();
                } else {
                    this.modalController.sendError(this.data || result);
                }
            }
        } catch (e) { }
        this.clicking = false;
    }

    ngOnInit() {
        let options = this.modalController.params;
        if (typeof options == "string") {
            options = {
                content: options,
            }
        }
        this.deal(options);
        this.form = this.formBuilder.group({
            data: [""],
        })
    }

    deal(options: ConfirmOptions) {
        this.content = options.content;
        this.content = options.content;
        this.buttonsCss = options.buttonsCss;
        this.contentTpl = options.contentTpl;
        if (!options.buttons) {
            options.buttons = [];
            if (!isBoolean(options.hideCancel) || !options.hideCancel) {
                options.buttons.push({
                    /**取消 */
                    text: options.cancelText || this.i18n.get("__3"),
                    cssClass: "btn-default-outline",
                    canClose: options.beforeCancel,
                    type: "cancel",
                } as any)
            }
            if (!isBoolean(options.hideOk) || !options.hideOk) {
                options.buttons.push({
                    /**确定 */
                    text: options.okText || this.i18n.get("__2"),
                    cssClass: "btn-success",
                    canClose: options.beforeOk,
                    type: "ok",
                } as any)
            }
        }
        this.buttons = options.buttons;
        if (options.inputs) {
            this.data = {};
            this.inputs = options.inputs;
            options.inputs.map(item => {
                this.data[item.name] = item.value || "";
                item.placeholder = item.placeholder || "";
                item.type = item.type || "text";
            })
        }
    }

    ngOnDestroy() { }
}
