// ng-modules/src/components/tabs/tab-content/tab-content.ts
import { Directive, ViewContainerRef, TemplateRef, Renderer2, ElementRef, Input } from "@angular/core";

@Directive({
	selector: "[tabContent]",
})
export class TabContentDirective {
	public inited: boolean;
	public openStatus: boolean = false;
	constructor(
		public viewContainer: ViewContainerRef,
		public templateRef: TemplateRef<any>,
		public renderer: Renderer2,
		public elementRef: ElementRef,
	) { }
	@Input("tabContent") public name: string;

	/**获取根节点 */
	getRootElement(): HTMLElement {
		let view = this.viewContainer.get(0);
		let element: HTMLElement = (<any>view).rootNodes[0];
		return element;
	}

	/**打开 */
	onopen(isopen?: boolean) { }
	open() {
		if (!this.openStatus) {
			if (!this.inited) {
				this.viewContainer.createEmbeddedView(this.templateRef)
			} else {
				let element: HTMLElement = this.getRootElement();
				element && this.renderer.removeClass(element, "hide");
				this.viewContainer.get(0).reattach();
			}
			this.openStatus = true;
			this.onopen();
			this.inited = true;
		}
	}

	/**关闭 */
	onclose() { }
	close() {
		if (this.openStatus) {
			if (this.inited) {
				let element: HTMLElement = this.getRootElement();
				element && this.renderer.addClass(element, "hide");
				this.viewContainer.get(0).detach();
			}
			this.openStatus = false;
			this.onclose();
		}
	}

}
