// ng-modules/src/components/form/daterange/daterange.ts
import { Directive, ElementRef, forwardRef } from "@angular/core";
import { ControlValueAccessor, NG_VALUE_ACCESSOR } from "@angular/forms";
import { Input } from "@angular/core";
import { UtilProvider } from "../../../providers/util/util";
import { I18nProvider } from "../../../providers/i18n/i18n";

@Directive({
	selector: "[daterange]",
	providers: [
		{ provide: NG_VALUE_ACCESSOR, useExisting: forwardRef(() => DaterangeDirective), multi: true }
	],
	// template: "<ng-content></ng-content>",
})
export class DaterangeDirective implements ControlValueAccessor {
	writeValue(value: any) {
		let format = "YYYY-MM-DD"
		if (this.daterange && this.daterange.locale && this.daterange.locale.format) {
			format = this.daterange.locale.format
		}
		if (value && value.match(" - ")) {
			let moment = (<any>window).moment;
			let data = value.split(" - ");
			$(this.element.nativeElement).data("daterangepicker").setStartDate(new Date(data[0]));
			$(this.element.nativeElement).data("daterangepicker").setStartDate(new Date(data[1]));
			$(this.element.nativeElement).val(moment(data[0]).format(format) + " - " + moment(data[1]).format(format));
		} else {
			$(this.element.nativeElement).val("");
		}
	}
	_registerOnChange() { }
	registerOnChange(fn: any) {
		this._registerOnChange = fn;
	}
	touched() { }
	registerOnTouched(fn: any) {
		this.touched = fn;
	}
	setDisabledState?(isDisabled: boolean) { }
	constructor(
		private element: ElementRef,
		private util: UtilProvider,
	) { }

	_config;
	@Input("daterange") set daterange(value) {
		this._config = value;
		this.setting(value);
	};
	get daterange() {
		return this._config;
	}

	ngOnInit() {
	}

	setting(config?: any) {
		let data = this.util.extend({
			singleDatePicker: false,
			autoUpdateInput: false,
			locale: I18nProvider.get("dateSetting"),
		}, config, true);
		$(this.element.nativeElement).daterangepicker(data);
		let that = this;
		$(this.element.nativeElement).on('apply.daterangepicker', function (ev, picker) {
			let format = "YYYY-MM-DD"
			if (that.daterange && that.daterange.locale && that.daterange.locale.format) {
				format = that.daterange.locale.format
			}
			that._registerOnChange();
			$(this).val(picker.startDate.format(format) + " - " + picker.endDate.format(format));
		});

		$(this.element.nativeElement).on('cancel.daterangepicker', function () {
			// $(this).val('');
		});
	}

}
