// ng-modules/src/components/tabs/tab-switch/tab-switch.ts
import { Directive, Input, Renderer2 } from "@angular/core";
import { Output } from "@angular/core";
import { ElementRef } from "@angular/core";
import { TabContentDirective } from "./../tab-content/tab-content";
import { EventEmitter } from "@angular/core";

@Directive({
	selector: "[tabSwitch]",
	host: {
		"(click)": "onclick()",
	}
})
export class TabSwitchDirective {
	content: TabContentDirective = null;

	constructor(
		public elementRef: ElementRef,
		public renderer: Renderer2,
	) { }

	@Output() enter = new EventEmitter();
	@Output() leave = new EventEmitter();

	@Input("tabSwitch") public name: string;
	/**是否默认打开 */
	_checked: boolean = false;
	@Input("checked") public set checked(checked) {
		if (typeof checked == "string") {
			this._checked = true;
		} else if (checked) {
			this._checked = true;
		} else {
			this._checked = false;
		}
	};
	public get checked() {
		return this._checked;
	}

	/**点击之前处理方法 */
	beforeClick() { }

	/**点击时 */
	onclick() {
		if (!this.content.openStatus) {
			this.beforeClick();
			if (this.content) {
				this.content.open();
			}
			this.checked = true;
			/**添加打开状态样式 */
			this.renderer.addClass(this.elementRef.nativeElement, "active");
		}
	}
}
