// ng-modules/test/pages/list/list.ts
import { Component } from "@angular/core";
import { ModalProvider, PageEvent, ListComponent as LC, ListTdData } from "../../../src/index";
import { ViewChild } from "@angular/core";

@Component({
	templateUrl: "./list.html",
	selector: "test-list",
	styleUrls: ["./list.scss"],
})
export class ListComponent {
	@ViewChild("treelist") treelist: LC;
	@ViewChild("list") list: LC;
	tree: any = [
		{
			id: 1, name: "文件", size: "40M", children: [
				{ id: 11, name: "北海旅游.avi", size: "30M" },
				{
					id: 12, name: "北海旅游图库", size: "10M", children: [
						{ id: 121, name: "TK001.jpg", size: "2M" },
						{ id: 122, name: "TK002.jpg", size: "2M" },
						{ id: 123, name: "TK003.jpg", size: "2M" },
						{ id: 124, name: "TK004.jpg", size: "2M" },
						{ id: 125, name: "TK005.jpg", size: "2M" },
					]
				},
			],
		},
		{
			id: 2, name: "资料", size: "1Gb", children: [
				{
					id: 21, name: "入党资料", size: "500M", children: [
						{ id: 211, name: "党史资料", size: "250M" },
						{ id: 212, name: "培训资料", size: "250M" },
					]
				}
			]
		}
	];

	constructor(
		private model: ModalProvider,
	) { }

	ngOnInit() {
		console.log(this)
		this.list.setCalcRowspan(this.calcRowspan)
	}

	getTree(e: PageEvent) {
		e.list = this.tree;
		e.totalSize = 0;
		e.complete(e);
	}
	saveStatus() {
		let status = this.treelist.treetableInstance.getStatus(this.treelist.list, (item) => item.id);
		console.log(status)
	}
	recoveryStatus() {
		this.treelist.treetableInstance.recoveryStatus(this.treelist.list, item => item.id, ["1_1", "12_2"])
	}

	getList(e: PageEvent) {
		e.list = [
			{ name: "ActiveStatus", value: 1, descr: "启用状态" },
			{ name: "ActiveStatus", value: 2, descr: "启用状态" },
			{ name: "DataType", value: 0, descr: "数据类型" },
			{ name: "DataType", value: 1, descr: "数据类型" },
			{ name: "DataType", value: 2, descr: "数据类型" },
			{ name: "DataType", value: 3, descr: "数据类型" },
		]
		e.totalSize = 0;
		e.complete(e);
	}
	calcRowspan(e: ListTdData, ele: HTMLElement) {
		let prop = e.prop;
		if (prop == "name" || prop == "descr") {
			if ((!e.pre) || (e.pre[prop] != e.current[prop])) {
				let rowspan = 1;
				while (e.list[e.rowIndex + rowspan] && (e.current[prop] == e.list[e.rowIndex + rowspan][prop])) {
					rowspan++;
				}
				e.setRowspan(rowspan);
			} else if (e.pre[prop] == e.current[prop]) {
				ele.style.display = "none"
			} else {
				e.setRowspan(1);
			}
		}
	}


}
