// ng-modules/test/pages/form/form.ts
import { Component } from "@angular/core";
import { ModalProvider, FormValidators } from "../../../src/index";
import { FormGroup } from "@angular/forms";
import { FormBuilder } from "@angular/forms";
import { Validators } from "@angular/forms";

@Component({
	templateUrl: "./form.html",
	selector: "test-form",
})
export class FormComponent {

	editForm: FormGroup;
	constructor(
		private model: ModalProvider,
		private formBuilder: FormBuilder,
	) { }

	date = {
		datetime: "",
		daterange: "",
	}

	ngOnInit() {
		this.editForm = this.formBuilder.group({
			name: ["", Validators.compose([Validators.required])],
			phoneNo: ["", Validators.compose([Validators.required, FormValidators.isPhoneNo("手机号码不正确")])],
			gender: ["", Validators.required],
			datetime: [""],
			daterange: [""],
		})
		console.log(this);
	}

	submit() {
		let error = FormValidators.checkError(this.editForm.controls)
		if (error) {
			this.model.toast({
				content: `${error}`,
				position: "ct",
				type: "error",
			})
		}
	}
}
