// ng-modules/test/pages/modal/modal.ts
import { Component, ChangeDetectorRef } from "@angular/core";
import { ModalProvider } from "../../../src/index";

@Component({
	templateUrl: "./modal.html",
	selector: "test-modal",
})
export class ModalComponent {

	constructor(
		private model: ModalProvider,
		private cdr: ChangeDetectorRef,
	) { }

	date;

	ngOnInit() {
		setInterval(() => {
			this.date = "" + ((+new Date) + 1);
			this.cdr.detectChanges();
		}, 1000)
	}

	alert(tmp?: any) {
		console.log(tmp)
		if (tmp) {
			return this.model.alert({
				contentTpl: tmp,
			})
		}
		this.model.alert("你好，angular！")
	}

	alertAutoClose() {
		this.model.alert({
			content: "3s后将会自动关闭",
			timeout: 3000,
		})
	}

	confirm(tpl?) {
		this.model.confirm({
			title: "温馨提示",
			content: "是否确认这个弹框？是否确认这个弹框？是否确认这个弹框？是否确认这个弹框？是否确认这个弹框？是否确认这个弹框？是否确认这个弹框？是否确认这个弹框？是否确认这个弹框？是否确认这个弹框？是否确认这个弹框？是否确认这个弹框？是否确认这个弹框？是否确认这个弹框？是否确认这个弹框？是否确认这个弹框？是否确认这个弹框？是否确认这个弹框？是否确认这个弹框？是否确认这个弹框？是否确认这个弹框？是否确认这个弹框？是否确认这个弹框？是否确认这个弹框？是否确认这个弹框？是否确认这个弹框？是否确认这个弹框？是否确认这个弹框？是否确认这个弹框？是否确认这个弹框？是否确认这个弹框？是否确认这个弹框？是否确认这个弹框？",
			contentTpl: tpl,
			inputs: [
				{ name: "name", label: "姓名", },
				{ name: "age", label: "年龄", },
				{
					name: "role", label: "角色", type: "select", options: [
						{ value: 1, name: "学生" },
						{ value: 2, name: "老师" },
					],
				}
			]
		}).then(data => console.log("ok", data), e => console.log("cancel", e))
	}

	infoConfirm(tpl?) {
		this.model.inforConfirm({
			content: "咋地了",
			contentTpl: tpl,
		})
	}

	loading() {
		let loading = this.model.loading("加载中，3s后关闭");
		setTimeout(() => {
			loading.destroy();
		}, 3000);
	}

	toast() {
		this.model.toast("你好，世界，我是一个信息提示框，希望您能接收到这条重要信息！")
	}

	toastWithClose() {
		this.model.toast({
			content: "这是一个信息提示框，希望您能接收到这条重要信息！需要点击关闭才能关闭！",
			timeout: 999999,
		})
	}

	toastWarning(type) {
		this.model.toast({
			content: "这是一条信息，你已经看到啦，提示 type=" + type,
			type: type,
		})
	}

	toastPosition(position) {
		this.model.toast({
			content: "这是一个关于方位的信息提示框，希望你可以看到这条消息，哈哈，当前提示框位置为 position=" + position + " timestamp=" + (+new Date),
			position: position,
			timeout: 9999,
		})
	}
}
