// ng-modules/dist/components/modal/toast/toast.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Component, ViewChildren, QueryList } from "@angular/core";
import { AnimationProvider } from "../../../providers/animation/animation";
import { ChangeDetectorRef } from "@angular/core";
var ToastComponent = (function () {
    function ToastComponent(animationProvider, cdr) {
        this.animationProvider = animationProvider;
        this.cdr = cdr;
        this.text = '';
        this.rightBC = [];
        this.rightTC = [];
        this.centerTC = [];
        this.centerCC = [];
        this.centerBC = [];
        this.leftTC = [];
        this.leftBC = [];
    }
    ToastComponent_1 = ToastComponent;
    ToastComponent.prototype.ngOnInit = function () {
        ToastComponent_1.Instance = this;
    };
    ToastComponent.prototype.calcMsgType = function (option) {
        var _a, _b, _c, _d, _e;
        switch (option.type) {
            case "success":
                _a = [option.backgroundColor || "#27AE60", option.color || "#fff", option.icon || "fa-thumbs-o-up"], option.backgroundColor = _a[0], option.color = _a[1], option.icon = _a[2];
                break;
            case "info":
                _b = [option.backgroundColor || "#3498DB", option.color || "#fff", option.icon || "fa-smile-o"], option.backgroundColor = _b[0], option.color = _b[1], option.icon = _b[2];
                break;
            case "error":
                _c = [option.backgroundColor || "#E74C3C", option.color || "#fff", option.icon || "fa-frown-o"], option.backgroundColor = _c[0], option.color = _c[1], option.icon = _c[2];
                break;
            case "warning":
                _d = [option.backgroundColor || "#D35400", option.color || "#fff", option.icon || "fa-meh-o"], option.backgroundColor = _d[0], option.color = _d[1], option.icon = _d[2];
                break;
            default:
                _e = [option.backgroundColor || "#27AE60", option.color || "#fff", option.icon || "fa-smile-o"], option.backgroundColor = _e[0], option.color = _e[1], option.icon = _e[2];
                break;
        }
    };
    ToastComponent.prototype.checkPosition = function (option) {
        var data = [
            [this.rightTC, "rt", -1, this.rtcDom],
            [this.rightBC, "rb", 1, this.rbcDom],
            [this.centerTC, "ct", -1, this.ctcDom],
            [this.centerCC, "cc", 0, this.cccDom],
            [this.centerBC, "cb", 1, this.cbcDom],
            [this.leftTC, "lt", -1, this.ltcDom],
            [this.leftBC, "lb", 1, this.lbcDom],
        ];
        var positionData;
        data.some(function (item) {
            if (option.position == item[1]) {
                positionData = item;
                return true;
            }
        });
        return positionData;
    };
    ToastComponent.prototype.calcPosition = function (option, list, type, direction, domList) {
        var _this = this;
        option.style = {
            'background-color': option.backgroundColor,
            'color': option.color,
        };
        var index;
        if (direction == -1) {
            index = 0;
            list.unshift(option);
        }
        else if (direction == 1) {
            index = list.push(option) - 1;
        }
        var subscribe = domList.changes.subscribe(function () {
            return setTimeout(function () {
                subscribe.unsubscribe();
                var dom = domList.toArray()[index];
                option.dom = dom;
                var contentDom = dom.nativeElement.getElementsByClassName("content")[0];
                if (type == "ct" || type == "cb") {
                    var contentWidth = contentDom.clientWidth;
                    if (contentWidth > 650) {
                        contentDom.style.width = '650px';
                    }
                    option.style.width = contentDom.clientWidth + 40 + 24 + 'px';
                    _this.animationProvider.fadeIn(dom.nativeElement);
                }
                var contentHeight = contentDom.clientHeight;
                option.height = contentHeight + 10;
                option.ready = true;
                option.style.height = option.height + "px";
                if (direction == 1) {
                    for (var i = index - 1; i >= 0; i--) {
                        var o_option = list[i];
                        var bottom = parseFloat(o_option.style.bottom || 0) + option.height + 16;
                        o_option.style.bottom = bottom + "px";
                    }
                }
                else if (direction == -1) {
                    for (var i = index + 1; i < list.length; i++) {
                        var o_option = list[i];
                        var top_1 = parseFloat(o_option.style.top || 0) + option.height + 16;
                        o_option.style.top = top_1 + "px";
                    }
                }
                option.close = function () {
                    clearTimeout(option.timeoutFn);
                    if (direction == 1) {
                        for (var i = 0;; i++) {
                            var o_option = list[i];
                            var bottom = parseFloat(o_option.style.bottom || 0) - option.height - 16;
                            o_option.style.bottom = bottom + "px";
                            if (o_option == option) {
                                list.splice(i, 1);
                                break;
                            }
                        }
                    }
                    else if (direction == -1) {
                        for (var i = list.length - 1;; i--) {
                            var o_option = list[i];
                            var top_2 = parseFloat(o_option.style.top || 0) - option.height - 16;
                            o_option.style.top = top_2 + "px";
                            if (o_option == option) {
                                list.splice(i, 1);
                                break;
                            }
                        }
                    }
                };
                option.timeoutFn = setTimeout(function () {
                    option.close();
                }, (option.timeout * 1000 || 0));
                _this.cdr.detectChanges();
            }, 0);
        });
    };
    ToastComponent.prototype.create = function (option) {
        if (!option.content)
            return;
        if (!option.position)
            option.position = "rb";
        this.calcMsgType(option);
        if (typeof option.timeout == "undefined")
            option.timeout = 3;
        var data = this.checkPosition(option);
        data.unshift(option);
        this.calcPosition.apply(this, data);
    };
    var ToastComponent_1;
    ToastComponent.Instance = null;
    __decorate([
        ViewChildren("rbcDom"),
        __metadata("design:type", QueryList)
    ], ToastComponent.prototype, "rbcDom", void 0);
    __decorate([
        ViewChildren("rtcDom"),
        __metadata("design:type", QueryList)
    ], ToastComponent.prototype, "rtcDom", void 0);
    __decorate([
        ViewChildren("ctcDom"),
        __metadata("design:type", QueryList)
    ], ToastComponent.prototype, "ctcDom", void 0);
    __decorate([
        ViewChildren("cccDom"),
        __metadata("design:type", QueryList)
    ], ToastComponent.prototype, "cccDom", void 0);
    __decorate([
        ViewChildren("cbcDom"),
        __metadata("design:type", QueryList)
    ], ToastComponent.prototype, "cbcDom", void 0);
    __decorate([
        ViewChildren("ltcDom"),
        __metadata("design:type", QueryList)
    ], ToastComponent.prototype, "ltcDom", void 0);
    __decorate([
        ViewChildren("lbcDom"),
        __metadata("design:type", QueryList)
    ], ToastComponent.prototype, "lbcDom", void 0);
    ToastComponent = ToastComponent_1 = __decorate([
        Component({
            template: "<div class=\"toast-container\"> <!-- \u53F3\u4E0B\u89D2 --> <div class=\"toast-area-right-bottom\" *ngIf=\"rightBC.length\"> <div class=\"toast-card\" [class.has-icon]=\"item.icon\" #rbcDom *ngFor=\"let item of rightBC\" [class.show]=\"item.ready\" [ngStyle]=\"item.style\"> <div [class]=\"'toast-icon fa '+item.icon\"></div> <div class=\"content\" [innerHTML]=\"item.content\"></div> <div class=\"toast-close\" [title]=\"'__4' | I18N\" (click)=\"item.close()\">\u00D7</div> </div> </div> <!-- \u53F3\u4E0B\u89D2\u7ED3\u675F --> <!-- \u53F3\u4E0A\u89D2 --> <div class=\"toast-area-right-top\" *ngIf=\"rightTC.length\"> <div class=\"toast-card\" [class.has-icon]=\"item.icon\" #rtcDom *ngFor=\"let item of rightTC\" [class.show]=\"item.ready\" [ngStyle]=\"item.style\"> <div [class]=\"'toast-icon fa '+item.icon\"></div> <div class=\"content\" [innerHTML]=\"item.content\"></div> <div class=\"toast-close\" [title]=\"'__4' | I18N\" (click)=\"item.close()\">\u00D7</div> </div> </div> <!-- \u53F3\u4E0A\u89D2\u7ED3\u675F --> <!-- \u5DE6\u4E0B\u89D2 --> <div class=\"toast-area-left-bottom\" *ngIf=\"leftBC.length\"> <div class=\"toast-card\" [class.has-icon]=\"item.icon\" #lbcDom *ngFor=\"let item of leftBC\" [class.show]=\"item.ready\" [ngStyle]=\"item.style\"> <div [class]=\"'toast-icon fa '+item.icon\"></div> <div class=\"content\" [innerHTML]=\"item.content\"></div> <div class=\"toast-close\" [title]=\"'__4' | I18N\" (click)=\"item.close()\">\u00D7</div> </div> </div> <!-- \u5DE6\u4E0B\u89D2\u7ED3\u675F --> <!-- \u5DE6\u4E0A\u89D2 --> <div class=\"toast-area-left-top\" *ngIf=\"leftTC.length\"> <div class=\"toast-card\" [class.has-icon]=\"item.icon\" #ltcDom *ngFor=\"let item of leftTC\" [class.show]=\"item.ready\" [ngStyle]=\"item.style\"> <div [class]=\"'toast-icon fa '+item.icon\"></div> <div class=\"content\" [innerHTML]=\"item.content\"></div> <div class=\"toast-close\" [title]=\"'__4' | I18N\" (click)=\"item.close()\">\u00D7</div> </div> </div> <!-- \u5DE6\u4E0A\u89D2\u7ED3\u675F --> <!-- \u4E2D\u4E0A --> <div class=\"toast-area-center-top\" *ngIf=\"centerTC.length\"> <div class=\"toast-card\" [class.has-icon]=\"item.icon\" #ctcDom *ngFor=\"let item of centerTC\" [class.show]=\"item.ready\" [ngStyle]=\"item.style\"> <div [class]=\"'toast-icon fa '+item.icon\"></div> <div class=\"content\" [innerHTML]=\"item.content\"></div> <div class=\"toast-close\" [title]=\"'__4' | I18N\" (click)=\"item.close()\">\u00D7</div> </div> </div> <!-- \u4E2D\u4E0A --> <!-- \u4E2D\u4E0B --> <div class=\"toast-area-center-bottom\" *ngIf=\"centerBC.length\"> <div class=\"toast-card\" [class.has-icon]=\"item.icon\" #cbcDom *ngFor=\"let item of centerBC\" [class.show]=\"item.ready\" [ngStyle]=\"item.style\"> <div [class]=\"'toast-icon fa '+item.icon\"></div> <div class=\"content\" [innerHTML]=\"item.content\"></div> <div class=\"toast-close\" [title]=\"'__4' | I18N\" (click)=\"item.close()\">\u00D7</div> </div> </div> <!-- \u4E2D\u4E0B --> </div> ",
            selector: "toast-view",
        }),
        __metadata("design:paramtypes", [AnimationProvider,
            ChangeDetectorRef])
    ], ToastComponent);
    return ToastComponent;
}());
export { ToastComponent };
