// ng-modules/dist/components/ueditor/ueditor.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Component, ViewChild, Input, ElementRef, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
var UEditorComponent = (function () {
    function UEditorComponent() {
        this._registerOnChange = function (data) { };
        this.isInit = false;
        this.ueReady = false;
    }
    UEditorComponent_1 = UEditorComponent;
    UEditorComponent.prototype.writeValue = function (obj) {
        var _this = this;
        this._defaultContent = obj;
        if (this.ue && this.ueReady) {
            setTimeout(function () {
                if (_this.isInit) {
                    return;
                }
                _this.isInit = true;
                _this.ue.setContent(obj);
            }, 200);
        }
    };
    UEditorComponent.prototype.registerOnChange = function (fn) {
        this._registerOnChange = fn;
    };
    UEditorComponent.prototype.registerOnTouched = function (fn) {
    };
    UEditorComponent.prototype.setDisabledState = function (isDisabled) {
    };
    UEditorComponent.prototype.ngOnInit = function () {
        var _this = this;
        if (!window.UE) {
            return console.error("请先引入ueditor,详情请联系 ikrong.com");
        }
        for (var key in window.UE.instants) {
            window.UE.instants[key].destroy();
            delete window.UE.instants[key];
        }
        window.UE.delEditor(this.editor.nativeElement);
        this.ue = window.UE.getEditor(this.editor.nativeElement, Object.assign({
            toolbars: [["paragraph", "fontfamily", "fontsize", "|", 'forecolor', 'backcolor', 'bold', 'italic', 'underline', 'strikethrough', "lineheight", 'superscript', 'subscript', "|", "indent", "justifyleft", "justifycenter", "justifyright", "justifyjustify", "|", "link", "unlink", "|", 'removeformat', 'formatmatch', 'autotypeset', 'blockquote', 'pasteplain', "|", "insertorderedlist", "insertunorderedlist", "rowspacingtop", "rowspacingbottom", "directionalityltr", "directionalityrtl", "touppercase", "pagebreak", "spechars", "|", "simpleupload", "imagenone", "imageleft", "imageright", "imagecenter", "|", "selectall", "cleardoc", "source", "fullscreen"]],
            autoHeightEnabled: false,
        }, this.config));
        this.ue.ready(function () {
            var height = _this.editor.nativeElement.offsetParent.clientHeight - _this.editor.nativeElement.querySelector(".edui-editor-toolbarbox").clientHeight;
            _this.ue.setHeight(height < 0 ? 400 : height);
            _this.ueReady = true;
            setTimeout(function () {
                if (_this.isInit) {
                    return;
                }
                _this.isInit = true;
                _this.ue.setContent(_this._defaultContent || "");
            }, 200);
            _this.ue.on("contentChange", function () {
                _this._registerOnChange(_this.ue.getContent());
            });
        });
    };
    UEditorComponent.prototype.getContent = function () {
        var data = this.ue.getContent();
        this._registerOnChange(data);
        return data;
    };
    var UEditorComponent_1;
    __decorate([
        ViewChild("editor"),
        __metadata("design:type", ElementRef)
    ], UEditorComponent.prototype, "editor", void 0);
    __decorate([
        Input("config"),
        __metadata("design:type", Object)
    ], UEditorComponent.prototype, "config", void 0);
    UEditorComponent = UEditorComponent_1 = __decorate([
        Component({
            selector: 'ueditor',
            template: "<div #editor class=\"editor-content\"></div> ",
            styles: [".editor-content {\n  width: 100%;\n  height: 100%;\n}"],
            providers: [
                {
                    provide: NG_VALUE_ACCESSOR,
                    useExisting: forwardRef(function () { return UEditorComponent_1; }),
                    multi: true
                }
            ]
        }),
        __metadata("design:paramtypes", [])
    ], UEditorComponent);
    return UEditorComponent;
}());
export { UEditorComponent };
