// ng-modules/dist/components/modal/alert/alert.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Component } from "@angular/core";
import { ModalController } from "../../../providers/modal/modalController";
var AlertComponent = (function () {
    function AlertComponent(modalController) {
        this.modalController = modalController;
        this.image = "";
        this.content = "";
    }
    AlertComponent.prototype.ngOnInit = function () {
        var options = this.modalController.params;
        if (typeof options == "string") {
            options = {
                content: options,
            };
        }
        this.deal(options);
    };
    AlertComponent.prototype.deal = function (options) {
        this.content = options.content;
        this.image = options.image;
        this.contentTpl = options.contentTpl;
    };
    AlertComponent.prototype.ngOnDestroy = function () { };
    AlertComponent = __decorate([
        Component({
            template: "<div class=\"alert-container\"> <div class=\"alert-image\" *ngIf=\"image&&(!contentTpl)\"> <img [src]=\"image\"> </div> <div class=\"alert-content\" *ngIf=\"(!contentTpl)\" [innerHTML]=\"content\"></div> <div *ngIf=\"contentTpl\"> <ng-container *ngTemplateOutlet=\"contentTpl\"> </ng-container> </div> </div> ",
        }),
        __metadata("design:paramtypes", [ModalController])
    ], AlertComponent);
    return AlertComponent;
}());
export { AlertComponent };
