// ng-modules/src/components/fullscreen/fullscreen.ts
import { Component, ElementRef } from '@angular/core';

@Component({
    selector: "fullscreen",
	templateUrl: "./fullscreen.html",
})
export class FullscreenComponent {
    distance = 10;
    widthTimestamp = 0;
    constructor(
        private elementRef: ElementRef,
    ) { }

    getRate() {
        let element: HTMLElement = this.elementRef.nativeElement
        let parentScroll = element.offsetParent.scrollTop || window.scrollY;
        let height = this.height;
        let offsetTop = element.offsetTop;
        let cut = offsetTop - parentScroll;
        return {
            rate: cut / height,
            cut: cut,
            height: height,
            offsetTop: offsetTop,
        };
    }

    ngOnInit() {
        let element: HTMLDivElement = this.elementRef.nativeElement
        let timer;
        element.addEventListener("mousewheel", () => {
            timer && clearTimeout(timer);
            timer = setTimeout(() => {
                let rate = this.getRate();
                if (rate.rate < 0.2 && rate.rate > 0) {
                    // $(element.offsetParent).animate({
                    //     scrollTop: rate.offsetTop,
                    // }, 500)
                }
            }, 200);
        }, {
                passive: true,
            } as any);
    }

    _width;
    get width() {
        if ((+new Date) - this.widthTimestamp > this.distance) {
            this.widthTimestamp = (+new Date)
            return this._width = window.innerWidth
        } else {
            return this._width;
        }
    }
    heightTimestamp = 0;
    _height;
    get height() {
        if ((+new Date) - this.heightTimestamp > this.distance) {
            this.heightTimestamp = (+new Date)
            return this._height = window.innerHeight
        } else {
            return this._height;
        }
    }
}
